langchain/chains/combine_documents/reduce.py

```diff
@@ -0,0 +1,276 @@
+from langchain.chains.combine_documents.base import BaseCombineDocumentsChain
+from langchain.docstore.document import Document
+
+
+class CombineDocsProtocol(Protocol):
+    """Interface for the combine_docs method."""
+
+    def __call__(self, docs: List[Document], **kwargs: Any) -> str:
+        """Interface for the combine_docs method."""
+
+
+class AsyncCombineDocsProtocol(Protocol):
+    """Interface for the combine_docs method."""
+
+    async def __call__(self, docs: List[Document], **kwargs: Any) -> str:
+        """Async nterface for the combine_docs method."""
+
+
+def _split_list_of_docs(
+    docs: List[Document], length_func: Callable, token_max: int, **kwargs: Any
+) -> List[List[Document]]:
+    new_result_doc_list = []
+    _sub_result_docs = []
+    for doc in docs:
+        _sub_result_docs.append(doc)
+        _num_tokens = length_func(_sub_result_docs, **kwargs)
+        if _num_tokens > token_max:
+            if len(_sub_result_docs) == 1:
+                raise ValueError(
+                    "A single document was longer than the context length,"
+                    " we cannot handle this."
+                )
+            new_result_doc_list.append(_sub_result_docs[:-1])
+            _sub_result_docs = _sub_result_docs[-1:]
+    new_result_doc_list.append(_sub_result_docs)
+    return new_result_doc_list
+
+
+def _collapse_docs(
+    docs: List[Document],
+    combine_document_func: CombineDocsProtocol,
+    **kwargs: Any,
+) -> Document:
+    result = combine_document_func(docs, **kwargs)
+    combined_metadata = {k: str(v) for k, v in docs[0].metadata.items()}
+    for doc in docs[1:]:
+        for k, v in doc.metadata.items():
+            if k in combined_metadata:
+                combined_metadata[k] += f", {v}"
+            else:
+                combined_metadata[k] = str(v)
+    return Document(page_content=result, metadata=combined_metadata)
+
+
+async def _acollapse_docs(
+    docs: List[Document],
+    combine_document_func: AsyncCombineDocsProtocol,
+    **kwargs: Any,
+) -> Document:
+    result = await combine_document_func(docs, **kwargs)
+    combined_metadata = {k: str(v) for k, v in docs[0].metadata.items()}
+    for doc in docs[1:]:
+        for k, v in doc.metadata.items():
+            if k in combined_metadata:
+                combined_metadata[k] += f", {v}"
+            else:
+                combined_metadata[k] = str(v)
+    return Document(page_content=result, metadata=combined_metadata)
+
+
+class ReduceDocumentsChain(BaseCombineDocumentsChain):
+    """Combining documents by recursively reducing them.
+
+    This involves
+    - combine_document_chain
+    - collapse_document_chain
+
+    `combine_document_chain` is ALWAYS provided. This is final chain that is called.
+    We pass all previous results to this chain, and the output of this chain is
+    returned as a final result.
+
+    `collapse_document_chain` is used if the documents passed in are too many to all
```

Comment:
nit: `collapse_document_chain` is used if there are more documents than can be passed to `combine_document_chain` in a single call.